fix(ngEventDirs): check for $rootScope.$$phase in event handler and don't $apply if already in $digest

Narretz opened the pull request:
<!-- General PR submission guidelines https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#submit-pr -->
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**



**What is the current behavior? (You can also link to an open issue here)**



**What is the new behavior (if this is a feature change)?**



**Does this PR introduce a breaking change?**



**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

markgardner committed:
fix(ngEventDirs): don't wrap the event handler in $apply if already in $digest

Digest cycle already in progress error can inadvertently be caused when triggering an
element's click event while within an active digest cycle. This is due to the ngEventsDirs
event handler always calling $rootScope.$apply regardless of the status of $rootScope.$$phase.
Checking the phase and calling the function immediately if within an active digest cycle
will prevent the problem without reducing current functionality.

Closes #14673
Closes #14674

Narretz committed:
fix(ngEventDirs): pass error in handler to $exceptionHandler when event was triggered in a digest

This ensures that the error handling is the same for events triggered inside and outside a digest.

petebacondarwin reviewed (approved):
Except for that surprising test, this LGTM

jbedard commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -149,5 +149,75 @@ describe('event directives', function() {
+        $rootScope.$apply(function() {
+          element.triggerHandler('click');
+        });
+      }).toThrowError('listener error');
```
I thought this error would be caught in the [`$apply` phase try/catch](https://github.com/Narretz/angular.js/blob/61c8910d7547a691f9d6f63129d0e16ffb4f93db/src/ng/rootScope.js#L1188) 🤔 

Also I think we could use another test which invokes the handler within a watcher, so it would be [within the `$digest` phase](https://github.com/Narretz/angular.js/blob/61c8910d7547a691f9d6f63129d0e16ffb4f93db/src/ng/rootScope.js#L849)

But note that in both of these cases it does not re-throw, so why does it in this test?

gkalpak replied:
I haven't confirmed, but I suspect it is because `$exceptionHandler` throws by default in tests. In order to test the "real" behavior, you have to set `$exceptionHandler` in `log` mode.

jbedard replied:
Ah yes, that sounds right. I guess we should have tests for both cases?

gkalpak commented on src/ng/directive/ngEventDirs.js:
```diff
@@ -64,10 +64,13 @@ forEach(
                 scope.$evalAsync(callback);
               } else {
-                scope.$apply(callback);
+                callback();
```
So, we are not wrapping this in `try { ... } catch (err) { $exceptionHandler(Err); }`?

jbedard replied:
That should make the following more consistent...
```
dom.click();
otherStuff();
```
... when `dom.click()` invokes `ng-click` and does vs does-not throw. SGTM...

Narretz commented:
I've updated the PR and added the additional try catch ... however, I think it's not actually necessary. See the two new tests: when we trigger an event handler in a watch expression fn, the error thrown gets logged twice because the watch loop in the digest never sees the error. 

If we simply throw in the watch expression fn, the watch loop in the digest catches the error and logs it (only once).

jbedard commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,35 +148,131 @@ describe('event directives', function() {
         });
-      }).toThrowError('listener error');
-    }));
+        $rootScope.do();
```
I was thinking this would be one line up, within the `$apply`. Or maybe it would be more clear if we did:
```
$rootScope.do = function() {
    element.triggerHandler('click');
    $log.log('done');
}
```
Then invoke `$rootScope.do` within vs  outside `$apply`. Those 2 cases should be consistent, both with `$exceptionHandler` throwing (the `done` never runs) vs only logging (the `done` always runs). I think this commit is required for those to be consistent?

Narretz replied:
Ok, if we do 
```
      $rootScope.do = function() {
          element.triggerHandler('click');
          $log.log('done');
        };

        $rootScope.$apply(function() {
          $rootScope.do();
        });
```
then we need the try-catch. 
I guess the question is now for which behavior  it's more important to be consistent? Because the try catch changes the behavior when we throw inside a digest. I assume that behavior is not as important.

jbedard replied:
Can you show an example of that change?

Narretz commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,6 +148,163 @@ describe('event directives', function() {
+    });
+
+
+    it('should not stop the digest when the event is fired in a watch expression fn', function() {
```
@jbedard if you trigger a click event handler inside a watch expression fn, then the additional try catch changes how the error gets logged, and in this case, anything that comes after the handler is called, whereas when you call the handler directly and throw, what comes after the handler is not called. See the test above for the different outcome.

jbedard replied:
That is caused by the `click()` vs `triggerHandler('click')`? I didn't think that would change anything...?

Narretz replied:
It's caused by the new try catch. To clarify, ij the watcher expression fn tests I test direct call to scope fn against event handler call to scope fn. The first test doesn't use the event handler at all. It's a bit of a digression from the original testing but it's an example where the new try catch leads to different behavior.

Narretz replied:
Is the watch expression fn test set up like you intended it to be? Because you mentioned that you think such a test would make sense.

jbedard replied:
Oh... I understand these two tests now. Sorry I was misreading `$rootScope.click()` as `$element.click()` for some reason (my fault 😅). I think this difference is ok, definitely worth the two `do()` tests behaving the same imo.

jbedard replied:
I think this test is correct and good as-is 👍

jbedard commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-describe('event directives', function() {
+fdescribe('event directives', function() {
```
This is fine for now, but making a note of it...

jbedard commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,6 +148,163 @@ describe('event directives', function() {
+      });
+    });
+
+    it('should not stop the digest when the handler is called in a watch expression fn', function() {
```
I don't think this one is necessary since it doesn't actually trigger a DOM event

gkalpak replied:
How is this test (and the ones below) test that the digest is not stopped? I would expect a second watch expression.
(What am I missing?)

jbedard commented:
LGTM

Could potentially also test `dom.click()` (the native `click` method) in addition to the the jQuery version. I think they should behave the same.

gkalpak commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,6 +148,163 @@ describe('event directives', function() {
+    });
+
+
+    it('should not stop the digest when the handler throws an error', function() {
```
This description seems wrong.

Narretz replied:
So what are we actually testing? "it should pass an error in the handler to the $exceptionHandler if the handler is called during a digest"?

gkalpak commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,6 +148,163 @@ describe('event directives', function() {
+    });
+  });
+
+  describe('click', function() {
```
Do we have similar test for `forceAsyncEvents`? If not, we should :grin:

Narretz commented on test/ng/directive/ngEventDirsSpec.js:
```diff
@@ -148,6 +148,163 @@ describe('event directives', function() {
+        };
+
+        $rootScope.$watch(function() {
+          $rootScope.click();
```
The test exists because I was looking for something that would behave differently when adding the try ... catch in the event directive.
